astroid/nodes/node_classes.py

```diff
@@ -5344,71 +5347,59 @@ def postinit(self, *, patterns: list[Pattern]) -> None:
+    bool: Const,
+    int: Const,
+    float: Const,
+    complex: Const,
+    str: Const,
+    bytes: Const,
 }
 
 
-def _update_const_classes():
-    """update constant classes, so the keys of CONST_CLS can be reused"""
-    klasses = (bool, int, float, complex, str, bytes)
-    for kls in klasses:
-        CONST_CLS[kls] = Const
-
-
-_update_const_classes()
-
-
-def _two_step_initialization(cls, value):
+def _two_step_initialization(
+    cls: type[_ContainerNodesT], value: list[Any]
+) -> _ContainerNodesT:
     instance = cls()
     instance.postinit(value)
-    return instance
+    return instance  # type: ignore[return-value]
 
 
-def _dict_initialization(cls, value):
-    if isinstance(value, dict):
-        value = tuple(value.items())
+def _dict_initialization(cls: type[Dict], value: list[Any]) -> Dict:
     return _two_step_initialization(cls, value)
 
 
-_CONST_CLS_CONSTRUCTORS = {
-    List: _two_step_initialization,
-    Tuple: _two_step_initialization,
-    Dict: _dict_initialization,
-    Set: _two_step_initialization,
-    Const: lambda cls, value: cls(value),
-}
-
-
-def const_factory(value):
-    """return an astroid node for a python value"""
-    # XXX we should probably be stricter here and only consider stuff in
-    # CONST_CLS or do better treatment: in case where value is not in CONST_CLS,
-    # we should rather recall the builder on this value than returning an empty
-    # node (another option being that const_factory shouldn't be called with something
-    # not in CONST_CLS)
+def const_factory(value: Any) -> List | Set | Tuple | Dict | Const | EmptyNode:
+    """Return an astroid node for a python value."""
     assert not isinstance(value, NodeNG)
 
-    # Hack for ignoring elements of a sequence
+    # This only handles instances of the CONST types. Any
+    # subclasses get inferred as EmptyNode.
+    # TODO: See if we should revisit these with the normal builder.
+    if value.__class__ not in CONST_CLS:
+        node = EmptyNode()
+        node.object = value
+        return node
+
+    # TODO: We pass an empty list as elements for a sequence
     # or a mapping, in order to avoid transforming
     # each element to an AST. This is fixed in 2.0
     # and this approach is a temporary hack.
+    if isinstance(value, list):
+        return _two_step_initialization(List, [])
+    if isinstance(value, set):
+        return _two_step_initialization(Set, [])
+    if isinstance(value, tuple):
+        return _two_step_initialization(Tuple, [])
     if isinstance(value, (list, set, tuple, dict)):
-        elts = []
-    else:
-        elts = value
-
-    try:
-        initializer_cls = CONST_CLS[value.__class__]
-        initializer = _CONST_CLS_CONSTRUCTORS[initializer_cls]
-        return initializer(initializer_cls, elts)
-    except (KeyError, AttributeError):
-        node = EmptyNode()
-        node.object = value
-        return node
+        return _dict_initialization(Dict, [])
+    return Const(value)
```

Comment:
We do get a mapping from `CONST_CLS`. Maybe we could use that to speed up the if block here. Would something like this work?
```py
    initializer_cls = CONST_CLS[value.__class__]
    if isinstance(initializer_cls, (List, Set, Tuple)):
        return _two_step_initialization(initializer_cls, [])
    if isinstance(initializer_cls, Dict):
        return _dict_initialization(initializer_cls, [])
    return Const(value)
```

Reply:
I added this in the last commit but used `issubclass` as we're comparing classes. This doesn't pass `pyright` though and I don't really understand why. `mypy` is also having trouble with it.
Could you perhaps have a quick look locally?

Reply:
> I added this in the last commit but used `issubclass` as we're comparing classes.

👍🏻 


Revert the TypeVar change, i.e. use `bound=Union["List", "Set, "Tuple", "Dict"]`. That will at least fix the pyright errors. Mypy somehow tries to find a common ancestor in `BaseContainer` and passes that along. Probably a bug.

Reply:
This also required re-adding the initial ignore. Let me know what you think!

Reply:
> This also required re-adding the initial ignore. Let me know what you think!

Not ideal, but that seems to be the best possible typing.